### PR TITLE
Wire up react router to the primary navigation

### PR DESCRIPTION
**Pull request by ian-leggett**
Since we have built all the collection list pages in React we now have an opportunity to wire up React router which will initially deal with the primary routes into the application (collection list pages), modules such as Companies, Events, Interactions...

To do this we need to take control of the Data hub header, to do this we will now have two versions in the code base, one is a complete copy from the npm package and the other is a full React version. We need two as collections pages are now FULL page React apart from Investments and the services such as Find exporters.

Things to focus on in this PR are:

**1. Data Hub header - React version** 

The React version uses React router and serves the following collection pages

1. Companies
2. Contacts
3. Events
4. Interactions
5. Orders

Investments needs a bit more React work as it includes tabs, adding this to the PR will just bloat what is already a massive PR.

Each link should have their own query parameters appended to them so that their default state is set every time you visit these pages. When cycling through these links the page will not reload as they are all served client side using react router (SPA behaviour).

**2. Data Hub header - Nunjucks**

All pages apart from the collection lists(minus Investments) will use the old header as the rest of the application is served via express and not routed client side. I have copied the code that is packaged via NPM and added it back in the codebase (header.njk). Everything thing is the same apart from the query params which I added to the links that point to the collection list pages as again they need the default state set.

**3. Breadcrumbs**

Whenever the collection list link labels(Companies, interactions…) appear as links within the breadcrumbs we need to make sure that the default query params are again set on the link. Most of the time this will happen as they are generated via the urls module and middleware however in some cases the link is passed down which overrides the breadcrumbs, if this happens there is some logic to append the params to the end (I have also updated unit tests).

**4. Title tags**

On the full React collection list pages the title tags are generated via the useEffect hook, see - defaultLayout

**5. Permissions**

Links - As we are now using a full React Data Hub header depending on permissions we need to hide/show links from the user. In order for us to know what permissions a user has we are still using express.js and injecting those permissions into the store via the provider.jsx I have created a component called ProtectedLink which handles this functionality and is used in the navigation.

Content - If the user tries to access a collection list in which they do not have permission, then they should see a 403 page as per pages served by express.js. I have created a component called ProtectedRoute that handles this functionality.

Note the actual mechanism that allows a user to access content comes from SSO so the work in React is just around creating the user experience the secure part is between express and the back end (server side). 

In a nutshell the app should behave as normal apart from the when you cycle through React routes as the browser will not reload.

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied.
https://github.com/uktrade/data-hub-frontend/blob/master/docs/Code%20review%20guidelines.md"

- [x] Has the branch been rebased to master?
- [x] Automated tests (Any of the following when applicable: Unit, Functional or End-to-End)
- [x] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/client/provider.jsx [hunk 1]
```diff
@@ -106,16 +106,18 @@ import {
   CONTACTS_LIST_ID,
   COMPANY_CONTACTS_LIST_ID,
 } from '../apps/contacts/client/state'
-import contactsReducer from '../apps/contacts/client/reducer'
 
-import { ID as INTERACTIONS_ID } from '../apps/interactions/client/state'
-import interactionsReducer from '../apps/interactions/client/reducer'
+// import { ID as CONTACTS_ID } from './modules/Contacts/CollectionList/state'
```
Comment (tomkingston) [inline src/client/provider.jsx:110]: Should this commented line be deleted?

**Review by tomkingston — APPROVED** [review]

**Comment by tomkingston** [comment]
Forgot to put in the PR comments, but it's probably worth squashing up all the "Fix tests" commits iinto a single commit

File: src/client/provider.jsx (hunk 1 above)
Comment (ian-leggett) [reply src/client/provider.jsx]: Good spot, all sorted. I realised that I was still using the contacts collection list from the apps dir instead of modules. Commit -

File: src/client/routes.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import ContactsCollectionList from './modules/Contacts/CollectionList'
+import EventsCollectionList from './modules/Events/CollectionList'
+import InteractionsCollectionList from './modules/Interactions/CollectionList'
+import OmisCollectionList from './modules/Omis/CollectionList'
+
+const routes = {
+  companies: [
+    {
+      path: '/companies',
+      permittedModule: 'datahub:companies',
```
[the remark this answers is not in the document]
Comment (ian-leggett) [reply src/client/routes.js]: I agree this makes sense, I have updated this in the links as well as the routes so both components `ProtectedLink` and `ProtectedRoute` take in `module` as props rather than `permittedModule`

File: src/apps/contacts/router.js [hunk 3]
```diff
@@ -57,7 +57,7 @@ router.get('/:contactId/audit', getAudit)
 router.get('/:contactId/documents', renderDocuments)
 
 router.use(
-  urls.contacts.interactions.index.route,
+  '/:contactId/interactions',
```
[the remark this answers is not in the document]
Comment (ian-leggett) [reply src/apps/contacts/router.js:60]: I feel like we are going to move away from the urls module at some point especially when it comes to routes set in express as they will eventually disappear. Once contacts is all built in React this will be deleted.

File: src/apps/routers.js [hunk 4]
```diff
@@ -10,6 +10,27 @@ const subApps = fs.readdirSync(__dirname, { withFileTypes: true })
 
 const appsRouters = []
 
+// TODO: Add Investments
```
[the remark this answers is not in the document]
Comment (ian-leggett) [reply src/apps/routers.js]: Removed

File: src/templates/_components/header.njk [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    href: ( datahubDomain + 'support' ),
+    active: ( 'datahub-support' === params.active )
+} %}
+{%- set apps = [
+	{
+		permittedKey: datahubCrmKey,
+		activeKey: companiesKey,
+		canShow: ( companiesKey in crmApps ) if crmApps else true,
+		name: 'Companies',
+		href: ( datahubDomain + 'companies?archived[0]=false&sortby=modified_on:desc&page=1' )
```
[the remark this answers is not in the document]
Comment (ian-leggett) [reply src/templates/_components/header.njk]: Yeah this makes sense. I have updated the NPM package and removed this 👍

**Review by paulgain — APPROVED** [review]
